[12.0][ADD] website_sale_stock_provisioning_date: new module.

ernestotejeda opened the pull request:
Show the future provisioning date closest to the current date for a product in the eCommerce.

![2020-09-22_18-46](https://user-images.githubusercontent.com/38267832/93944964-fe377d00-fd03-11ea-8085-64a469025567.png)


cc @Tecnativa TT24985

pedrobaeza commented on website_sale_stock_provisioning_date/models/product_product.py:
```diff
@@ -0,0 +1,23 @@
+
+    def _get_next_provisioning_date(self):
+        domain = [
+            ("company_id", "=", self.env.user.company_id.id),
```
Is this the same way as Odoo uses in the stock info popup for sales?

ernestotejeda replied:
In the popup, the date that is shown is the expected date to deliver the products defined in the sale order.

![2020-09-22_11-17](https://user-images.githubusercontent.com/38267832/93902303-66b33980-fcc5-11ea-9b67-c950617d0258.png)

pedrobaeza replied:
Uhm, how about the expected one, how is computed?

ernestotejeda replied:
But the expected date is information related to the sales order and what we want is information related to a product. In this case, the nearest future date the product is expected to be received. In the same sale order. In the same sales order we could have several products with different expected reception dates.

pedrobaeza replied:
Yeah, but this computation is done taking the minimum expected date for each product. How that expected date is computed? We want the same algorithm (removing the customer lead time).

ernestotejeda replied:
That is in v13 and as I see in the code it is almos the same as the date of the sale order.
In v12 there is effective_date, but what it calculates is the minimum date of the done pickings related to the sale_order.
 I don't think those fields are useful for what we want.

pedrobaeza replied:
OK

pedrobaeza commented on website_sale_stock_provisioning_date/models/product_product.py:
```diff
@@ -0,0 +1,23 @@
+                .sudo()
+                .search(domain, order="date_expected", limit=1)
+        )
+        return move.date_expected
```
Show only date. Time part is mostly irrelevant.
```suggestion
        return move.date_expected.date()
```

pedrobaeza reviewed (commented):
Tested on runbot. Seems OK except the comments.

pedrobaeza commented on website_sale_stock_provisioning_date/models/product_product.py:
```diff
@@ -0,0 +1,23 @@
+                .sudo()
+                .search(domain, order="date_expected", limit=1)
+        )
+        return move and move.date_expected.date() or False
```
```suggestion
        return move.date_expected and move.date_expected.date() or False
```

HaraldPanten commented on website_sale_stock_provisioning_date/README.rst:
```diff
@@ -0,0 +1,93 @@
+Usage
+=====
+
+To use this module, you need to:
+
+#. Go to *Sales > Products > Products*, create a new product or edit an
+   existing one.
+#. Check 'Show Next Provisioning Date' field under eCommerce tab.
+#. Go to Website Shop.
+#. Click on a product that have 'Show Next Provisioning Date' checked.
+#. You will see the future provisioning date closest to the current date
+   for this product under the 'Add to Cart' button.
+
```
Just a picky comment: I would just add that it doesn't work with all of the "Availability" options that user can set in Ecommerce tab (in product form). It will work only with the ones that show the product quantity.

ernestotejeda replied:
I also think that it would not make much sense to show that date if the stock availability is not also shown. But we don't want to force this functionality to be used only for specific availability options. I have modified the field help so that whoever is going to use this field takes this into account.

HaraldPanten reviewed (approved):
Functional review 👍

Commit:
[ADD] website_sale_stock_provisioning_date: new module.

Show the future provisioning date closest to the current date
for a product in the eCommerce.

pedrobaeza commented:
/ocabot merge nobump
